### ISSUE TITLE
Return partial matrix results when some pairs error

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#3898]
We recently had a few reports of people who'd like the matrix to return partial results, in the case where some individual pairs throw (mostly because of exceeding max distance). That's a fair point and indeed pretty annoying.

I looked into that quickly to see what's needed. We currently check the `max_matrix_distance` during loki's `matrix_action` and then throw if one pair is exceeding that. To achieve partial results we'd either have to mark the combinations of sources & targets which exceed the max distance or pull that check into the matrix's `SourceToTargets` functions. But for the many:many CostMatrix it's not quite as easy as just ignoring a location in either/both sets (in 1:many that'd be more natural). For CostMatrix I think it's the best/easiest way to check distance/cost during the expansion. 

We already pass the max distance to the matrix algos to set the `current_cost_threshold_`, which we don't expand beyond. We could allow per request/config option to not check crow-distance on the locations in loki, but rely on the `current_cost_threshold_` check during the expansion in thor. The obvious disadvantage is that in worst case we'd expand all sources/targets until the `max_matrix_distance` is reached only to return `null` results. If that's what we'd like to do, I'd like to protect the server with a config option to disallow to return partial results and instead fail early (as now). For TimeDistanceMatrix we could still check for each origin the crow-distance to each target and exclude the ones which exceed, it's cheap enough.

It seems that's sorta the only pair-specific error we could choose to ignore (almost: we could also disallow not-connected pairs while calculating the matrix for connected pairs).

Any thoughts @dnesbitt61 @kevinkreiser ?

**Comment by nilsnolde** [comment]
One thing that seems rather exotic, even right now: if I read the code correctly, it seems we're even returning a "valid" result for pairs which exceeded a threshold (either the dynamic one we update during expansion or the static one derived from `max_matrix_distance`. So, a source/target combo which exceeds the threshold while expanding is still output as a result but not marked in any way. A user would think the provided time/distance is correct, but actually we never found the connection.

Shouldn't we change that in any case, to report `null` on such cases? (which might not happen too much currently, chances are high those cases are caught by loki and throw before they get to thor, but e.g. South Italy -> Albania, without ferries, should trigger that behavior)

**Comment by nilsnolde** [comment]
Uh, well, I was probably thinking of the wrong error :sweat_smile: @kevinkreiser thought about the case when one pair actually failed to find a path. Makes much more sense, we could do what I was talking about as well in theory, but it's more esoteric and I wouldn't follow that line of thought anymore for now.

**Comment by nilsnolde** [comment]
Although.. the original issue where this is coming from, does talk about getting over the `max_matrix_distance` error: https://github.com/valhalla/valhalla/discussions/3595#discussioncomment-2569412. 

Uff jeez, I really did quite a bit of work lately in the matrix code and I should've reacted more intuitively that I did: in fact @kevinkreiser , we already do what you suggested: when one pair fails to find a path (exceeding the cost threshold we determine for this request based on travel mode), we put `null` in the response JSON and don't fail the whole request. So that's great. 

But then it does turn out after all that people are annoyed by one or some pairs exceeding the `max_matrix_distance` and want the non-offending pairs to succeed. It's a bit of a cheeky request (I'm sure being unaware that it's a bit cheeky), we could argue that the client should make sure it's not exceeding the requirements set by the server (and communicated in the error body). It's not very hard for a client to implement a check like that, it's just the spherical distance of each pair. But it could also be implemented in the server, so that not every client has to check it individually. We might then introduce the problem that requests entirely out of range on all pairs would all expand to the point of the threshold and enough of those requests are a home-brewed DDOS attack (also things like bays/u-shaped geographies can easily lead there, but it's more rare I'd think). Even with the config/request option turned on, we could only allow for say 20% of pairs to exceed the `max_matrix_distance`, which sounds like a fair compromise.

In summary, I still think it could be a config (& maybe also request) option to allow to disable the distance check in loki and let the expansion deal with it by either finding the connection or not if the cost exceeds the threshold (which is still mainly set by the `max_matrix_distance` config option, so it's not going completely nuts on us). Sorry for the back and forth :smile:

**Comment by nilsnolde** [comment]
Ok so talking about this off GitHub, I was convinced that this is a cheap check every client can do. Just remember: it’s only spherical crow-distance between all pairs.

**Comment by michaelkrog** [comment]
@nilsnolde Do you have some tips for how to tackle it?

I understand that is it just a crow-distance calculation per pair. But if I pull a coordinate from the request I pull it for all pairs, right?

In a simple matrix we might have 3 locations; A, B , C with the following distances:

| Source | Target | Distance |
| ------- |------- |----------|
| A.        | B        | 300km |
| B.         | C      | 300km |
| A         | C        | 600km |

Let say `max_matrix_distance` is 400km.

- If I add all sources and targets it will fail. 
- If I remove A from sources it will not fail, but I won't get the distance between A & B either.
- If I remove C from target it will not fail, but I won't get the distance between B & C either.


Doing a crow-distance check is easy. But I can't seem to figure out how to then do a request giving me as much data as Valhalla can deliver for a list of SourceTarget pairs.

I noticed that the table service allows for a fallback crow-distance result for all the pairs that cannot be calculated
([See table docs](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/http.md#table-service)). This allows for retrieving all possible distances and then have the ones that fail still give a decent result.

**Comment by nilsnolde** [comment]
> This allows for retrieving all possible distances and then have the ones that fail still give a decent result.

personally I’m not so sure about “decent result”;) I’d rather have a “null” result so I know what’s going on. Let me think about it again, maybe there’s an easy way to support the “null” result scenario. Despite being pretty recent I tend to forget pitfalls quite easily:)
